postgres_copy/psycopg_compat.py

```diff
@@ -0,0 +1,43 @@
+try:
+    import psycopg  # noqa: F401  just detect the presence of psycopg(3)
+    from codecs import getincrementaldecoder
+    from io import TextIOBase
+
+    BUFFER_SIZE = 128 * 1024
+
+    class NoopDecoder:
+        def decode(self, input, final=False):
+            return input
+
+    utf8_decoder_cls = getincrementaldecoder("utf8")
+
+    def copy_to(cursor, sql, params, destination):
+        # psycopg2.copy_expert has special handling of encoding
+        if isinstance(destination, TextIOBase):
+            decoder = utf8_decoder_cls()
+        else:
+            decoder = NoopDecoder()
+        with cursor.copy(sql, params) as copy:
+            while data := copy.read():
+                data = decoder.decode(data)
+                destination.write(data)
+            # TODO: is this extra one needed?
+            if data := decoder.decode(b"", final=True):
+                destination.write(data)
```

Comment:
Can you explain this to me? I'm just ignorant of how the new version works.

Reply:
this is all related to the comment on line 15 `# psycopg2.copy_expert has special handling of encoding`

psycopg2 (but not 3) detects the destination being an "instance" of `TextIOBase` (https://github.com/psycopg/psycopg2/blob/c2b6a8aaeae71b30a96403bf5d6e7eafc21afaef/psycopg/utils.c#L217) and decodes the data before writing


the entire interface of for `copy` has change significantly in psycopg3. this pr tries to maintain the previous interface with all these hacks. maybe a better idea would be design a new interface that more closely matches the new `copy` https://www.psycopg.org/psycopg3/docs/basic/copy.html#copying-block-by-block
but it seems strange for this single library to have different interfaces for different psycopg versions and that makes migration harder for users